pkg/source/ttnv3/source.go

```diff
@@ -0,0 +1,206 @@
+		if s.config.networkServerGRPCAddress == "" {
+			logger.WithField("paths", nsPaths).Warn("Network Server disabled but fields specified to get")
+		} else {
+			ns, err := api.Dial(s.ctx, s.config.networkServerGRPCAddress)
+			if err != nil {
+				if !continueOnError {
+					return nil, err
+				}
+				logger.WithError(err).Error("Could not connect to Network Server")
+			} else {
+				nsRes, err := ttnpb.NewNsEndDeviceRegistryClient(ns).Get(s.ctx,
+					&ttnpb.GetEndDeviceRequest{
+						EndDeviceIds: ids,
+						FieldMask:    ttnpb.FieldMask(nsPaths...),
+					},
+				)
+				if err != nil {
+					if !continueOnError {
+						return nil, err
+					}
+					logger.WithError(err).Error("Could not get end device from Network Server")
+				} else {
+					if err := res.SetFields(nsRes, ttnpb.AllowedReachableBottomLevelFields(nsPaths, getEndDeviceFromNS, nsRes.FieldIsZero)...); err != nil {
+						return nil, err
+					}
+					updateDeviceTimestamps(&res, nsRes)
+				}
+			}
+		}
+	}
+	return &res, nil
+}
+
+// ExportDevice implements the source.Source interface.
+func (s Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
+	isPaths, nsPaths, asPaths, jsPaths := splitEndDeviceGetPaths(ttnpb.BottomLevelFields(ttnpb.EndDeviceFieldPathsNested)...)
+	if len(nsPaths) > 0 {
+		isPaths = append(isPaths, "network_server_address")
+	}
+	if len(asPaths) > 0 {
+		isPaths = append(isPaths, "application_server_address")
+	}
+	if len(jsPaths) > 0 {
+		isPaths = append(isPaths, "join_server_address")
+	}
+	is, err := api.Dial(s.ctx, s.config.identityServerGRPCAddress)
+	if err != nil {
+		return nil, err
+	}
+	ids := &ttnpb.EndDeviceIdentifiers{
+		ApplicationIds: &ttnpb.ApplicationIdentifiers{ApplicationId: s.config.appID},
+		DeviceId:       devID,
+	}
+	dev, err := ttnpb.NewEndDeviceRegistryClient(is).Get(s.ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIds: ids,
+		FieldMask:    ttnpb.FieldMask(isPaths...),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if dev.ClaimAuthenticationCode.GetValue() != "" {
+		// ClaimAuthenticationCode is already retrieved from the IS. We can unset the related JS paths
+		jsPaths = ttnpb.ExcludeFields(jsPaths, claimAuthenticationCodePaths...)
+	}
+	res, err := s.getEndDevice(ids, nsPaths, asPaths, jsPaths, true)
+	if err != nil {
+		return nil, err
+	}
+	paths := ttnpb.AddFields(nsPaths, ttnpb.AddFields(asPaths, ttnpb.AddFields(jsPaths, "ids.dev_addr")...)...)
+	if err := dev.SetFields(res, paths...); err != nil {
+		return nil, err
+	}
+	updateDeviceTimestamps(dev, res)
+	dev.ClaimAuthenticationCode = nil
+	return dev, nil
+}
+
+// RangeDevices implements the source.Source interface.
+func (s Source) RangeDevices(appID string, f func(source.Source, string) error) error {
+	is, err := api.Dial(s.ctx, s.config.identityServerGRPCAddress)
+	if err != nil {
+		return err
+	}
+	paths := ttnpb.RPCFieldMaskPaths["/ttn.lorawan.v3.EndDeviceRegistry/List"].Allowed
+	res, err := ttnpb.NewEndDeviceRegistryClient(is).List(
```

Comment:
Note that you will need pagination here - `List` will not return all of the end devices, and you need to use the `limit`/`page`/`order`.

Reply:
Fixed in https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/47/commits/fdc52a0a996ba1edaee8682574a866e3f64a9020